### ISSUE TITLE
Statement leak warning on start since 2.11.0.Final

### ISSUE THREAD

**Issue by famod** [quarkusio/quarkus#26995]
### Describe the bug

In all modes:
```
WARN  [io.agr.pool] (main) Datasource '<default>': JDBC resources leaked: 0 ResultSet(s) and 1 Statement(s)
```

### Expected behavior

No such warning (as in 2.10.3.Final)

### Actual behavior

Warning logged, with unknown source/reason

### How to Reproduce?

TODO

### Output of `uname -a` or `ver`

`Linux *** 5.10.102.1-microsoft-standard-WSL2 #1 SMP Wed Mar 2 00:30:59 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

### Output of `java -version`

`OpenJDK 64-Bit Server VM Zulu17.36+13-CA (build 17.0.4+8-LTS, mixed mode, sharing)`

### Quarkus version or git rev

2.11.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.6

### Additional information

```
Installed features: [agroal, blaze-persistence, cdi, config-yaml, hibernate-orm, hibernate-orm-panache, hibernate-validator, jacoco, jdbc-mariadb, liquibase, narayana-jta, qute, scheduler, smallrye-context-propagation, spring-data-jpa, spring-di, vertx]
```

I haven't digged deeper yet. IIRC, there are possibilities to get more output from agroal...

**Comment by yrodiere** [comment]
Not again :facepalm: 

I'll need a reproducer if you have one... This was supposed to be fixed in #14305, and we added a non-regression test so it supposedly still is fixed for the cases we knew about.

This might be a new "leak", either a real one or one that is wrongly reported by Agroal... that's possible if some operations are no longer happening strictly in the order Agroal expects.

**Comment by Sanne** [comment]
It could also relate to scope of context. I see `smallrye-context-propagation` among the list of extensions, and I'm wondering why - is there any reactive code being used?

**Comment by famod** [comment]
It's related to `quarkus-liquibase` and I suppose it's an upstream issue. I'll try the quickstart next...

**Comment by famod** [comment]
Errm... https://github.com/liquibase/liquibase/pull/2914/files#diff-b378dd1eeb35b9dbe8751bd7462c9aad1b1c8c74d6cd5fd6c64b4014d05a1260R53

I haven't bisected yet, but it's not happening with Liquibase 4.12 and the above change is in 4.13 and IMO it's a very hot candidate.

**Comment by famod** [comment]
Also, it makes sense now that I'm not seeing this in the quickstart because it doesn't import csv data via liquibase (`loadData`) but my project does.

**Comment by famod** [comment]
Yeah, it's that new static unclosed `ExecutablePreparedStatementBase.lastPreparedStatement`, verified via the following hack:
```java
        final var liquibaseFactory = getBean(LiquibaseFactory.class);
        try (Liquibase liquibase = liquibaseFactory.createLiquibase()) {
            liquibase.update(liquibaseFactory.createContexts(), new LabelExpression("dev"), false);
            // HACK STARTS HERE
            try {
                final var field = ExecutablePreparedStatementBase.class.getDeclaredField("lastPreparedStatement");
                field.setAccessible(true);
                final PreparedStatement stmt = (PreparedStatement) field.get(null);
                try {
                    if (stmt != null && !stmt.isClosed()) {
                        stmt.close();
                    }
                } catch (final SQLException e) {
                    throw new IllegalStateException(e);
                }
            } catch (final ReflectiveOperationException e) {
                throw new IllegalStateException(e);
            }
        }
```
(the above only works in a special `@QuarkusTest` of mine where I run Liquibase a second time for...reasons)

**Comment by famod** [comment]
@geoand we don't want to add such a nasty hack to the extension (for a limited time), do we?

**Comment by geoand** [comment]
Oh dear... That is very questionable choice on the liquibase side... If they really really had to do that, there should at least be a configuration option to force the closing of the statement...

I can't say I like the hack, but on the other hand the current experience is kinda bad for users because the warning is scary...

I'm on the fence...

**Comment by gsmet** [comment]
Hum. This is extremely questionable to say the least. They should at least expose a way to close the statement when we're done. But yeah, even with that, that wouldn't be pretty.

Given how bad it is, I could live with the hack. But we need to make sure it's property tested so that we catch it if they change the code and that we get it to work in native.

**Comment by Sanne** [comment]
N.B. Agroal actually ensures the statements are closed, so this is only about the warning.

Yes that caching layer in Liquibase is questionable, as statements can be cached at other layers too, e.g. the JDBC driver.

+1 to have temporary hack like @famod 's. I only hope it will be cleaned up eventually :)

An alternative would beto hide the warning during Liquibase processing, as long as we can re-enable it after liquibase is done. Do we have such a logging filtering capability?

**Comment by geoand** [comment]
We can have logging filters, but they don't come and go - they stick around and would need to know the state of things in order to properly know when to discard the warning

**Comment by famod** [comment]
I don't like that hack either but I don't see another short-term remedy. Maybe some other idea pops up in that liquibase issue but I won't hold my breath.

If we add that hack we shouldn't be throwing exceptions like that, IMO. Log a warning at most.

**Comment by Sanne** [comment]
actually I just remembered one thing: Agroal doesn't log by itself but it delegates such warnings to our `AgroalDataSourceListener`. So this coul be patched to become mutable and disable warnings for some time.

Not great, but probably less fragile - especially for native-image.

**Comment by Sanne** [comment]
This is the method to be patched:
 - https://github.com/quarkusio/quarkus/blob/9d5f9cb7ac4962b0ea29480df460233c6a2dad38/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalEventLoggingListener.java#L69-L71

**Comment by gsmet** [comment]
Question is: do we really want to keep a statement open?

**Comment by Sanne** [comment]
> Question is: do we really want to keep a statement open?

Agroal is closing them.

**Comment by famod** [comment]
> actually I just remembered one thing: Agroal doesn't log by itself but it delegates such warnings to our `AgroalDataSourceListener`.

Ah yes, that class looks familiar.

> So this coul be patched to become mutable and disable warnings for some time.

Hm, I can image something like `suppressNextWarningMatching("JDBC resources leaked: .* and 1 Statement")`, plus an explicit reset when the liquibase extension is done (so that it doesn't suppress anything unrelated).
Is run anything in parallel at this point (at startup)? I can see the warning is coming from the `main` thread.

**Comment by famod** [comment]
@Sanne WDYT about my proposal? (see previous comment)

**Comment by Sanne** [comment]
@famod yes that sounds great; could you prototype it?

**Comment by famod** [comment]
@Sanne I have very little time for that but I can try.

**Comment by Sanne** [comment]
sure - no pressure.

**Comment by geoand** [comment]
Is this still an issue with the latest version of Quarkus?

**Comment by famod** [comment]
liquibase/liquibase#3128 is still open and I'm still seeing it in 2.13.1.

**Comment by geoand** [comment]
Thanks for checking

**Comment by famod** [comment]
FTR, still an issue with Quarkus 3.2.2/Liquibase 4.20.0.

**Comment by wiesnery** [comment]
Still an issue with Quarkus 3.15.2, Liquibase 4.29.1 and agroal 2.5

Anyone still there who could fix it?
Would appreciate as our auditors are not very keen to see warnings around the db in the logs...

**Comment by yrodiere** [comment]
The leak does exist and it's in Liquibase, so there's not much we can do in Quarkus.

... except to stop reporting leaks, which would make auditors happy (though for all the wrong reasons). See https://github.com/quarkusio/quarkus/issues/44047 about that, which may require changes to Agroal.

The only other solution would be the [nasty reflection-based hack suggested above](https://github.com/quarkusio/quarkus/issues/26995#issuecomment-1198371051).

**Comment by yrodiere** [comment]
I'll consider this "fixed" by #44047
